lib/src/rollout_evaluator.dart

```diff
@@ -1,349 +1,1071 @@
-              setting.variationId,
-          value: rolloutRule?.value ?? percentageItem?.value ?? setting.value,
-          matchedEvaluationRule: rolloutRule,
-          matchedEvaluationPercentageRule: percentageItem);
+  EvaluationResult _evaluateSetting(Setting setting,
+      EvaluationContext evaluationContext, EvaluateLogger? evaluateLogger) {
+    EvaluationResult? evaluationResult;
+    if (setting.targetingRules.isNotEmpty) {
+      evaluationResult =
+          _evaluateTargetingRules(setting, evaluationContext, evaluateLogger);
+    }
+    if (evaluationResult == null && setting.percentageOptions.isNotEmpty) {
+      evaluationResult = _evaluatePercentageOptions(setting.percentageOptions,
+          setting.percentageAttribute, evaluationContext, null, evaluateLogger);
     }
+    evaluationResult ??= EvaluationResult(
+        variationId: setting.variationId,
+        value: setting.settingsValue,
+        matchedTargetingRule: null,
+        matchedPercentageOption: null);
 
-    try {
-      if (user == null) {
-        if (setting.rolloutRules.isNotEmpty ||
-            setting.percentageItems.isNotEmpty) {
-          _logger.warning(3001,
-              'Cannot evaluate targeting rules and % options for setting \'$key\' (User Object is missing). You should pass a User Object to the evaluation methods like `getValue()` in order to make targeting work properly. Read more: https://configcat.com/docs/advanced/user-object/');
+    return evaluationResult;
+  }
+
+  EvaluationResult? _evaluateTargetingRules(Setting setting,
+      EvaluationContext evaluationContext, EvaluateLogger? evaluateLogger) {
+    evaluateLogger?.logTargetingRules();
+    for (TargetingRule rule in setting.targetingRules) {
+      bool evaluateConditionsResult;
+      String? error;
+      try {
+        evaluateConditionsResult = _evaluateConditions(
+            rule.conditions,
+            rule,
+            evaluationContext,
+            setting.salt,
+            evaluationContext.key,
+            setting.segments,
+            evaluateLogger);
+      } on RolloutEvaluatorException catch (rolloutEvaluatorException) {
+        error = rolloutEvaluatorException.message;
+        evaluateConditionsResult = false;
+      }
+
+      if (!evaluateConditionsResult) {
+        if (error != null) {
+          evaluateLogger?.logTargetingRuleIgnored();
         }
-        logEntries.add('Returning ${setting.value}');
-        return produceResult();
+        continue;
+      }
+      if (rule.servedValue != null) {
+        return EvaluationResult(
+            variationId: rule.servedValue!.variationId,
+            value: rule.servedValue!.settingsValue,
+            matchedTargetingRule: rule,
+            matchedPercentageOption: null);
       }
 
-      logEntries.add('User object: $user');
+      if (rule.percentageOptions?.isEmpty ?? true) {
+        throw ArgumentError("Targeting rule THEN part is missing or invalid.");
```

Comment:
low: The semantics of `ArgumentError` doesn't really fit config model errors. [The Dart reference](https://api.dart.dev/stable/3.2.6/dart-core/ArgumentError-class.html) says:

> Error thrown when a function is passed an unacceptable argument.

So it would to check if Dart has an error type more appropriate to the situation. Or maybe it would be the best to define an custom type for config model errors because probably we'll do that in the future anyway. See https://github.com/configcat/.net-sdk/pull/86/files#diff-04fa3c4116eed9f3d337e2ebd0b40a92d57ad862146c92a03756674d9ba7ce08

Reply:
After talking over the review, this change will be added later.